### ISSUE TITLE
Default values no longer returned as response?

### ISSUE THREAD

**Issue by bstdenis** [geopython/pywps#226]
# Description

I am no longer getting the default values of inputs in request.inputs, see code below. Trying to access the inputs that are not provided crash the process. See code below and the error returned by the WPS process in the comments:

```
from pywps import Process,get_format,configuration
from pywps import LiteralInput,LiteralOutput,ComplexOutput

# Example usage:
#
#localhost/pywps?service=WPS&request=execute&version=1.0.0&identifier=test3&DataInputs=limit=2
#
# However, not providing the limit input fails:
#
#localhost/pywps?service=WPS&request=execute&version=1.0.0&#identifier=test3&DataInputs=
#
# Error:
# exceptionCode="NoApplicableCode"
# Process error: wps_test3.py._handler Line 42 'limit'

class Test3(Process):
    def __init__(self):
        inputs = [LiteralInput('limit',
                               'Some limit',
                               data_type='integer',
                               default=10,
                               min_occurs=0),]

        outputs = [LiteralOutput('test_result',
                                 'Test result',
                                 data_type='string')]

        super(Test3,self).__init__(
            self._handler,
            identifier='test3',
            title='test3 process',
            version='0.1',
            inputs=inputs,
            outputs=outputs,
            store_supported=True,
            status_supported=True)

    def _handler(self,request,response):
        # This fails with
        # exceptionCode="NoApplicableCode"
        # Process error: wps_test3.py._handler Line 42 'limit'
        limit = request.inputs['limit'][0].data
        response.outputs['test_result'].data = str(limit)
        
        # If we output the keys to request.inputs, we see that 'limit' is
        # not there (in this case it's an empty list):
        #response.outputs['test_result'].data = str(request.inputs.keys())

        # The only workaround I find currently is to fetch the default from
        # the initial list of inputs:
        #if 'limit' not in request.inputs:
        #    for one_input in self.inputs:
        #        if one_input.identifier == 'limit':
        #            default_str = str(one_input.default)
        #            response.outputs['test_result'].data = default_str
        #            break

        return response
```

# Environment
- operating system: Ubuntu 16.04
- Python version: 2.7.12
- PyWPS version: 4.0.0 (2017-01-13)
- source/distribution
  - [x] git clone
  - [ ] Debian
  - [ ] PyPI
  - [ ] zip/tar.gz
  - [ ] other (please specify): 
- web server
  - [x] Apache/mod_wsgi
  - [ ] CGI
  - [ ] other (please specify): 
# Steps to Reproduce
# Additional Information

**Comment by bstdenis** [comment]
I have a follow up question. Is pywps supposed to make the distinction between an input that was not provided at all and an input that was set to its default value? Here's a simple example where this matters:

Let's say I have a python function that has a "limit" parameter that's an integer, but that can also be "None" (for no limits). In WPS, since the LiteralInput is set to data_type integer, I believe it's default value must also be an integer. If the pywps "request" object tells me whether the limit input was provided or not, I could set it to None inside the _handler and proceed to call the python function that accepts either an integer or None. The other option is to add a boolean LiteralInput that determines whether my "limit" should be set or not. Does that make sense?

**Comment by kvold** [comment]
I have the same problem. If an optional input is not given it results in an KeyError in Apache (like in issue #154), and an "NoApplicableCode"-exception in the response. Should I provide more details?

**Comment by jachym** [comment]
Hi,

you are both right. The input parameter "default" was available there for LiteralInput, but it was not used. You could theoreticaly use it manually 

```
        limit = request.inputs['limit'][0].default
```

But that was not nice.

I've introduced more robust `default` value handler and it's now availabe for testing tin #236 pull request - could you check, if it works for you, please?

Thanks

**Comment by kvold** [comment]
Hey,

unfortunately it seems like I still get the same error after trying out your commit.:

56 [Mon Feb 20 09:57:29.939029 2017] [wsgi:error] [pid 9140] 2017-02-20 09:57:29,938 INFO sqlalchemy.engine.base.Engine COMMIT
257 [Mon Feb 20 09:57:29.953593 2017] [wsgi:error] [pid 9140] Traceback (most recent call last):
258 [Mon Feb 20 09:57:29.954230 2017] [wsgi:error] [pid 9140]   File "/usr/local/wps/src/pywps/pywps/app/Process.py", line 222, in     _run_process
259 [Mon Feb 20 09:57:29.954247 2017] [wsgi:error] [pid 9140]     wps_response = self.handler(wps_request, wps_response)
260 [Mon Feb 20 09:57:29.954559 2017] [wsgi:error] [pid 9140]   File "/usr/local/pywps/processes/elevationXML.py", line 56, in _han    dler
261 [Mon Feb 20 09:57:29.954575 2017] [wsgi:error] [pid 9140]     requestedNumberOfPoints = request.inputs['points'][0].data
262 [Mon Feb 20 09:57:29.954630 2017] [wsgi:error] [pid 9140] KeyError: 'points'


When I output the keys to request.inputs (as bstdenis did), the key "points" does not exist.

Here is the relevant process (the commented out try-except was just a hack to go around the problem in the first place):

    def __init__(self):
        inputs = [
            ComplexInput(identifier="gpx", title="GPX tracklog",
                         abstract="A tracklog in GPX format to be charted",
                         min_occurs=1, max_occurs=1,
                         supported_formats=[Format("application/gml+xml")],
                         mode=MODE.STRICT),
            LiteralInput(identifier="points", title="number of points",
                         abstract="Number of points to chart", min_occurs=0,
                         max_occurs=1, data_type='integer', default=1)]
        outputs = [
            ComplexOutput(identifier='xml', title='Enhanced track data',
                          abstract='Elevation and terrain',
                          supported_formats=[Format("application/xml")])

        super(ElevationXML, self).__init__(
            self._handler,
            identifier="elevationXML",  # the same as the file name
            title="elevationXML",
            version="0.2",
            inputs=inputs,
            outputs=outputs,
            store_supported=True,
            status_supported=True,
            abstract=self.__doc__,
        )

    def _handler(self, request, response):
        # Input
        gpxfile = request.inputs['gpx'][0].file
        #try:
        requestedNumberOfPoints = request.inputs['points'][0].data
        #except:
        #    requestedNumberOfPoints = 1

        # Run operations


And the request which fails:
http://localhost/wps?request=Execute&service=WPS&version=1.0.0&identifier=elevationXML&datainputs=gpx=@xlink:href=http://remotehost/elevation_readable.gpx

**Comment by bstdenis** [comment]
I second @kvold observation, the default value is not returned in the "request" object. Looking at the modifications you made in your branch, I understand that you are setting the data to the default value when the input object is created, but that does not trigger its inclusion in the "request" object.

There is a design decision to make here, is the "request" object strictly reserved for inputs that were provided by the user? If not, the unprovided inputs should be activated and set to their default value in the "request" object. Otherwise, we have to keep track of our inputs by name (instead of just creating an "inputs" list as we do above) and check for the presence of the key in the "request" object, e.g.:

`
class Test3(Process):
    def __init__(self):
        self.limit_input = LiteralInput('limit',
                               'Some limit',
                               data_type='integer',
                               default=10,
                               min_occurs=0)
        # potentially more inputs
        inputs = [self.limit_input,]

        outputs = [LiteralOutput('test_result',
                                 'Test result',
                                 data_type='string')]

        super(Test3,self).__init__(
            self._handler,
            identifier='test3',
            title='test3 process',
            version='0.1',
            inputs=inputs,
            outputs=outputs,
            store_supported=True,
            status_supported=True)

    def _handler(self,request,response):
        if 'limit' not in request.inputs:
            response.outputs['test_result'].data = str(self.limit_input.default)
        else:
            response.outputs['test_result'].data = str(request.inputs['limit'][0].data)

        return response
`

**Comment by jachym** [comment]
well, seems that this should be done along with min and max occurs attributes too :-/ sorry about it

**Comment by jachym** [comment]
I've just uploaded new version of the fix, in the https://github.com/jachym/pywps/tree/226_default_value branch. Will try to test it with provided process example from @bstdenis  but would appreciate, if someone had a look at it.

**Comment by bstdenis** [comment]
I tested your branch and the default value seems to work as expected, thx!

However, while playing around I noticed something else: if I set a default value and min_occurs/max_occurs, I can get some strange behaviours, e.g.:

```
inputs = [LiteralInput('limit',
                               'Some limit',
                               data_type='integer',
                               default=10,
                               min_occurs=2,
                               max_occurs=3),]
```

If I do not provide the 'limit' parameter in my wps call, I will get the default value (once) as the result, which is inconsistent with the min_occurs parameter as far as I can tell.
If I provide the limit parameter only one time (which is not enough because of the min_occurs), I get a

File "/usr/local/lib/python2.7/dist-packages/pywps-4.0.0-py2.7.egg/pywps/app/Service.py", line 503, in create_literal_inputs
[Wed Mar 29 19:22:35.526774 2017] [wsgi:error] [pid 23:tid 140134455895808] [remote 172.17.0.1:63125]     raise MissingParameterValue(locator=source.identifier)
[Wed Mar 29 19:22:35.526800 2017] [wsgi:error] [pid 23:tid 140134455895808] [remote 172.17.0.1:63125] TypeError: __init__() takes at least 2 arguments (2 given)

If I provide more than 2 'limit' parameter, I get the proper behaviour.

I must say though, I'm not really sure I would ever want to use the default value in combination with a parameter that also has min_occurs... just food for thoughts...
